src/encoder/encoderopus.cpp

```diff
@@ -0,0 +1,439 @@
+
+    if (result != OPUS_OK) {
+        kLogger.warning() << "opus_encoder_create failed:" << opusErrorString(result);
+        return -1;
+    }
+
+    // Optimize encoding for high-quality music
+    opus_encoder_ctl(m_pOpus, OPUS_SET_COMPLEXITY(10)); // Highest setting
+    opus_encoder_ctl(m_pOpus, OPUS_SET_SIGNAL(OPUS_SIGNAL_MUSIC));
+
+    if(m_bitrateMode == 0) {
+        // == Constrained VBR ==
+        // Default mode, gives the best quality/bitrate compromise
+        opus_encoder_ctl(m_pOpus, OPUS_SET_BITRATE(m_bitrate * 1000)); // convert to bits/second
+        opus_encoder_ctl(m_pOpus, OPUS_SET_VBR(1)); // default value in libopus
+        opus_encoder_ctl(m_pOpus, OPUS_SET_VBR_CONSTRAINT(1)); // Constrained VBR
+    } else if(m_bitrateMode == 1) {
+        // == CBR (quality loss at low bitrates) ==
+        opus_encoder_ctl(m_pOpus, OPUS_SET_BITRATE(m_bitrate * 1000)); // convert to bits/second
+        opus_encoder_ctl(m_pOpus, OPUS_SET_VBR(0));
+    } else if(m_bitrateMode == 2) {
+        // == Full VBR ==
+        opus_encoder_ctl(m_pOpus, OPUS_SET_VBR(1));
+        opus_encoder_ctl(m_pOpus, OPUS_SET_VBR_CONSTRAINT(0)); // Unconstrained VBR
+    }
+
+    // Size the input FIFO buffer with twice the maximum possible sample count that can be
+    // processed at once, to avoid skipping frames or waiting for the required sample count
+    // and encode at a regular pace.
+    // This is set to the buffer size of the sidechain engine because
+    // Recording (which uses this engine) sends more samples at once to the encoder than
+    // the Live Broadcasting implementation
+    m_pFifoBuffer = new FIFO<CSAMPLE>(EngineSideChain::SIDECHAIN_BUFFER_SIZE * 2);
+
+    double samplingPeriodMs = (1.0/((float)m_samplerate)) * 1000.0;
+    double samplesPerChannel = kOpusFrameMs / samplingPeriodMs;
+
+    m_readRequired = samplesPerChannel * m_channels;
+    m_pFifoChunkBuffer = new CSAMPLE[m_readRequired * sizeof(CSAMPLE)]();
+    initStream();
+
+    return 0;
+}
+
+void EncoderOpus::initStream() {
+    ogg_stream_clear(&m_oggStream);
+    ogg_stream_init(&m_oggStream, getSerial());
+    m_header_write = true;
+    m_granulePos = 0;
+    m_packetNumber = 0;
+
+    pushHeaderPacket();
+    pushTagsPacket();
+}
+
+// Binary header construction is done manually to properly
+// handle endianness of multi-byte number fields
+void EncoderOpus::pushHeaderPacket() {
+    // Opus identification header
+    // Format from https://tools.ietf.org/html/rfc7845.html#section-5.1
+
+    // Header buffer size:
+    // - Magic signature: 8 bytes
+    // - Version: 1 byte
+    // - Channel count: 1 byte
+    // - Pre-skip: 2 bytes
+    // - Samplerate: 4 bytes
+    // - Output Gain: 2 bytes
+    // - Mapping family: 1 byte
+    // - Channel mapping table: ignored
+    // Total: 19 bytes
+    const int frameSize = 19;
+    unsigned char* data = new unsigned char[frameSize]();
```

Comment:
Plain pointers are evil and the extra () is unnecessary. Why don't you use an array that is statically allocated on the stack? It's just 19 bytes.